src/server/common/resultsets/game_history.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+   Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY.
+
+   See the COPYING file for more details.
+*/
+
+#ifdef HAVE_MYSQLPP
+
+#include "server/common/resultsets/game_history.hpp"
+#include "serialization/string_utils.hpp"
+
+void game_history::read(mariadb::result_set_ref rslt)
+{
+    while(rslt->next())
+    {
+        result r;
+        r.game_name = rslt->get_string("GAME_NAME");
+        r.reload = rslt->get_boolean("RELOAD");
+        r.game_start = rslt->get_date_time("START_TIME").str();
+        r.scenario_name = rslt->get_string("SCENARIO_NAME");
+        r.scenario_id = rslt->get_string("SCENARIO_ID");
+        r.era_name = rslt->get_string("ERA_NAME");
+        r.era_id = rslt->get_string("ERA_ID");
+        for(const auto& player_info : utils::split(rslt->get_string("PLAYERS")))
+        {
+            std::vector<std::string> info = utils::split(player_info, ':');
+            if(info.size() == 2)
+            {
+                r.players.emplace_back(player{ info[0], info[1] });
+            }
+        }
+        r.modifications = utils::split(rslt->get_string("MODIFICATIONS"));
+        r.replay_url = rslt->get_string("REPLAY_URL");
+        results.emplace_back(r);
+    }
+}
+
+std::unique_ptr<simple_wml::document> game_history::to_doc()
+{
+    simple_wml::document doc;
+    for(const auto& result : results)
+    {
+        simple_wml::node& ghr = doc.root().add_child("game_history_result");
+        ghr.set_attr_dup("game_name", result.game_name.c_str());
+        ghr.set_attr_int("reload", result.reload);
+        ghr.set_attr_dup("game_start", result.game_start.c_str());
+        ghr.set_attr_dup("scenario_name", result.scenario_name.c_str());
+        ghr.set_attr_dup("scenario_id", result.scenario_id.c_str());
+        ghr.set_attr_dup("era_name", result.era_name.c_str());
+        ghr.set_attr_dup("era_id", result.era_id.c_str());
+        for(const auto& player : result.players)
+        {
+            simple_wml::node& p = ghr.add_child("player");
+            p.set_attr_dup("name", player.name.c_str());
+            p.set_attr_dup("faction", player.faction.c_str());
+        }
+        for(const auto& mod : result.modifications)
+        {
+            simple_wml::node& m = ghr.add_child("modification");
+            m.set_attr_dup("name", mod.c_str());
+        }
+    }
+    return doc.clone();
```

Comment:
How come we clone the doc here? Couldn't `doc` be created as a unique_ptr on the heap and just returned?

Reply:
Mostly just it's still kind of weird to be using a unique_ptr to be able to return something.  Updated.

Reply:
Hmm... You could make it return a `document` and add a move ctor/assignment op for when that function is actually used (since the `document` copy ctor is deleted).

Reply:
I'll leave that to someone else - this works, and I'm fine with not needing to try to figure out how to do that for simple_wml.